src/TokenParser/TypesTokenParser.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\TokenParser;
+
+use Twig\Error\SyntaxError;
+use Twig\ExpressionParser;
+use Twig\Node\Expression\ArrayExpression;
+use Twig\Node\Expression\ConstantExpression;
+use Twig\Node\Expression\NameExpression;
+use Twig\Node\Node;
+use Twig\Node\SetNode;
+use Twig\Node\TypesNode;
+use Twig\Token;
+use Twig\TokenStream;
+
+/**
+ * Declare variable types.
+ *
+ * The mapping must use names as keys (not strings) and type strings as values.
+ * 
+ * Names may be suffixed with `?` to mark them as optional (i.e., the context may not
+ * have that variable defined).
+ * 
+ * Example:
+ * ```
+ *    {% types {foo: 'int', bar?: 'string'} %}
+ * ```
+ * @author Jeroen Versteeg <jeroen@alisqi.com>
+ * @see    https://github.com/twigphp/Twig/issues/4165
+ * @internal
+ */
+final class TypesTokenParser extends AbstractTokenParser
+{
+    public function parse(Token $token): Node
+    {
+        $stream = $this->parser->getStream();
+
+        $expression = ExpressionParser::parseNameToStringMapping($stream, true);
```

Comment:
```suggestion
        $expression = $this->parser->getExpressionParser()->parseNameToStringMapping($stream, true);
```

Reply:
That would work, but the `TokenStream` argument could be completely different than `$this->parser`'s instance variable.

It's also not possible to create a `new ExpressionParser($this->parser, $env)` in `TypesTokenParser` because the latter doesn't have the required `Twig\Environment`. (`Parser->env` is private and there's no `public` getter.)

Is there a more elegant solution?

Reply:
But this is how it's done everywhere. I would follow that.

Reply:
Thanks for the suggestion.

Unfortunately, this doesn't work, because `Parser` only sets its `$expressionParser` in `parse()`. Once that method has returned, the stream has already been parsed. (There's no way to rewing the `TokenStream`.)

I did remove `static` by simply creating a `new ExpressionParser()`.